File: vibe-coding/todo-app/frontend_web/src/modules/lists_mod/application/search_lists/SearchListsDto.ts
/**
 * Entrada del caso de uso SearchLists. Gemelo de `search_lists_dto.py`.
 *
 * Misma forma que en Python: campos de solo lectura, constructor privado y una
 * factoria `fromPrimitives` que convierte lo que llega de la interfaz.
 */
export class SearchListsDto {
    private constructor(public readonly nameContains: string) {}

    public static fromPrimitives(primitives: Record<string, unknown>): SearchListsDto {
        return new SearchListsDto(String(primitives["nameContains"] ?? "").trim());
    }
}

File: vibe-coding/todo-app/backend_web/src/modules/lists_mod/application/search_lists/search_lists_dto.py
from dataclasses import dataclass
from typing import Any, Self, final

from src.modules.lists_mod.domain.enums.list_field_enum import ListFieldEnum


@final
@dataclass(frozen=True, slots=True)
class SearchListsDto:
    """Entrada del caso de uso SearchLists (filtros del listado).

    Un caso de uso de busqueda tambien tiene DTO de entrada, aunque los filtros
    sean opcionales. Es lo que evita que manana alguien anada un filtro leyendo
    directamente de la peticion dentro del service.
    """

    name_contains: str

    @classmethod
    def from_primitives(cls, primitives: dict[str, Any]) -> Self:
        # Vacio = sin filtro. El repositorio lo traduce a un LIKE '%%', que casa
        # con todo, de forma que hay una sola consulta y no dos caminos.
        return cls(name_contains=str(primitives.get(ListFieldEnum.NAME_CONTAINS, "")).strip())
